File: shortread_RNAseq/extract_motif_information.R
library("tidyverse")


process_motif_data <- function(data, motif_column, filter_motifs) {
  motif = unlist(strsplit(motif_column, ".", fixed = T))[1]
  data %>%
    separate_rows(motif_column, sep = '\\),' ) %>%
    separate(., motif_column, into = c(paste0("Number_Motif_", motif), paste0("Strand_", motif), paste0("Value_", motif_column)), sep = ",") %>%
    separate(., paste0("Number_Motif_", motif), into = c(paste0("Distance_", motif), paste0("Motif_", motif)), sep = "\\(") %>%
    filter(., get(paste0("Strand_", motif)) == Strand & !motif %in% filter_motifs)
}







	args <- commandArgs(trailingOnly = TRUE)

	if (length(args) != 1) {
  		stop("Usage: Rscript your_script.R <input_file>")
	}

	input_file <- args[1]

	name = substr(input_file, start = 1, stop = nchar(input_file)-4)
	print(name)
        data = read.delim(input_file, header = T)
       

        data = process_motif_data(data, "AAUAAA.Distance.From.Peak.sequence.strand.conservation.", c("AAAAAA", "TTTTTT") )
        data = process_motif_data(data, "UUGUUU.Distance.From.Peak.sequence.strand.conservation.", c("TTTTTT", "AAAAAA") )
        data = process_motif_data(data, "TGTA.Distance.From.Peak.sequence.strand.conservation.", c(""))
        data = process_motif_data(data, "YA.Distance.From.Peak.sequence.strand.conservation.", c(""))

        data$AAUAAA_Distance = as.numeric(data$AAUAAA_Distance)
        data$UUGUUU_Distance = as.numeric(data$UUGUUU_Distance)
        data$TGTA_Distance = as.numeric(data$TGTA_Distance)
        data$YA_Distance = as.numeric(data$YA_Distance)

	print("4")

        positive = data %>% filter(Strand == "+" & 131 > (AAUAAA_Distance - TGTA_Distance) & (AAUAAA_Distance - TGTA_Distance) > 59 & 28 > (UUGUUU_Distance - AAUAAA_Distance) & (UUGUUU_Distance - AAUAAA_Distance) > 5 & YA_Distance - AAUAAA_Distance > 0 )
        negative = data %>% filter(Strand == "-" & 131 > (TGTA_Distance - AAUAAA_Distance) & (TGTA_Distance - AAUAAA_Distance) > 59 & 20 > (UUGUUU_Distance - AAUAAA_Distance) & (UUGUUU_Distance - AAUAAA_Distance) > 5 & AAUAAA_Distance - YA_Distance > 0 )


        perfekte = rbind(positive, negative)
        positionen_perfekt = perfekte %>% select(Chr, Start, End, Strand) %>% distinct()
	print("5")
        write.table(positionen_perfekt, paste0(name, "_perfect_motifs.bed"), quote = F, col.names = F, row.names = F, sep = "\t")

File: shortread_RNAseq/extract_motif_information.py
import pandas as pd
import re
import sys

def process_motif_data(data, motif_column_index, filter_motifs):
    motif_column = data.columns[motif_column_index]
    motif = motif_column.split(" ")[0]
    print(motif)
    df_split = data[motif_column].str.split('\),', expand=True)
    df_split['original_row'] = data.index

    df_expanded = df_split.melt(id_vars=['original_row'], value_vars=list(range(df_split.shape[1] - 1)))
    df_expanded = df_expanded.dropna()[['original_row', 'value']]
    df_result = pd.merge(data, df_expanded, left_index=True, right_on='original_row')
    df_result = df_result.drop(motif_column, axis=1).rename(columns={'value': motif_column})
    df_result = df_result.reset_index(drop=True)



    df_split = df_result[motif_column].str.split('[(),]', expand=True)
    df_split.columns = [f'{motif}_Numeric', f'{motif}_Sequence', f'{motif}_Strand', f'{motif}_Value', f'{motif}_empty']
    df_split = df_split.drop([f'{motif}_Value', f'{motif}_empty'], axis=1)

    df_result = df_result.drop(motif_column, axis=1)
    df_result = pd.concat([df_result, df_split], axis=1)
    df_result = df_result.loc[~df_result[f'{motif}_Sequence'].isin(filter_motifs)]

    # Reset the index if needed
    df_result = df_result.reset_index(drop=True)    
    
    return df_result



args = sys.argv[1:]

if len(args) != 1:
    raise ValueError("Usage: python your_script.py <input_file>")

input_file = args[0]

name = input_file[:-28]
print(name)
data = pd.read_csv(input_file, sep='\t')

data = process_motif_data(data, 21, ["AAAAAA", "TTTTTT"])
print("1")

data = process_motif_data(data, 21, ["TTTTTT", "AAAAAA"])
print("2")

data = process_motif_data(data, 22, [])
print("3")

data = process_motif_data(data, 23, [])
print("4")




data["AAUAAA_Numeric"]  = pd.to_numeric(data["AAUAAA_Numeric"])
data["UUGUUU_Numeric"]  = pd.to_numeric(data["UUGUUU_Numeric"])
data["TGTA_Numeric"]    = pd.to_numeric(data["TGTA_Numeric"])
data["YA_Numeric"]      = pd.to_numeric(data["YA_Numeric"])

print("5")

positive_upstream   = data[(data["Strand"] == "+") & (41 > (data["YA_Numeric"] - data["TGTA_Numeric"])) & ((data["YA_Numeric"] - data["TGTA_Numeric"]) > 29) & (26 > (data["YA_Numeric"] - data["AAUAAA_Numeric"])) & ((data["YA_Numeric"] - data["AAUAAA_Numeric"]) > 14) & ((data["UUGUUU_Numeric"] - data["YA_Numeric"]) < 0) & ((data["AAUAAA_Numeric"] - data["UUGUUU_Numeric"]) < 0)]
positive_downstream = data[(data["Strand"] == "+") & (41 > (data["YA_Numeric"] - data["TGTA_Numeric"])) & ((data["YA_Numeric"] - data["TGTA_Numeric"]) > 29) & (26 > (data["YA_Numeric"] - data["AAUAAA_Numeric"])) & ((data["YA_Numeric"] - data["AAUAAA_Numeric"]) > 14) & (21 > (data["UUGUUU_Numeric"] - data["YA_Numeric"])) & ((data["UUGUUU_Numeric"] - data["YA_Numeric"]) > 9)]

negative_upstream   = data[(data["Strand"] == "-") & (41 > (data["TGTA_Numeric"] - data["YA_Numeric"])) & ((data["TGTA_Numeric"] - data["YA_Numeric"]) > 29) & (26 > (data["AAUAAA_Numeric"] - data["YA_Numeric"])) & ((data["AAUAAA_Numeric"] - data["YA_Numeric"]) > 14) & ((data["UUGUUU_Numeric"] - data["YA_Numeric"]) > 0) & ((data["AAUAAA_Numeric"] - data["UUGUUU_Numeric"]) > 0)]
negative_downstream = data[(data["Strand"] == "-") & (41 > (data["TGTA_Numeric"] - data["YA_Numeric"])) & ((data["TGTA_Numeric"] - data["YA_Numeric"]) > 29) & (26 > (data["AAUAAA_Numeric"] - data["YA_Numeric"])) & ((data["AAUAAA_Numeric"] - data["YA_Numeric"]) > 14) & (21 > (data["YA_Numeric"] - data["UUGUUU_Numeric"])) & ((data["YA_Numeric"] - data["UUGUUU_Numeric"]) > 9)]


perfekte = pd.concat([positive_upstream, positive_downstream, negative_downstream, negative_upstream ])
positionen_perfekt = perfekte[["Chr", "Start", "End", "Strand"]].drop_duplicates()
print("6")
positionen_perfekt.to_csv(f"{name}_perfect_motifs.bed", sep='\t', header=False, index=False)
